packages/babel-types/test/cloning.js

```diff
@@ -72,4 +72,53 @@ describe("cloneNode", function() {
       node.declarations[0].id.typeAnnotation,
     );
   });
+
+  it("should support shallow cloning without loc", function() {
+    const node = t.variableDeclaration("let", [
+      t.variableDeclarator({
+        ...t.identifier("value"),
+        typeAnnotation: t.anyTypeAnnotation(),
+      }),
+    ]);
+    node.loc = {};
+    const cloned = t.cloneNode(node, /* deep */ false, /* withoutLoc */ true);
+    expect(cloned.loc).toBeNull();
+  });
+
+  it("should support deep cloning without loc", function() {
+    const node = t.variableDeclaration("let", [
+      t.variableDeclarator({
+        ...t.identifier("value"),
+        typeAnnotation: t.anyTypeAnnotation(),
+      }),
+    ]);
+    node.loc = {};
+    node.declarations[0].id.loc = {};
+    const cloned = t.cloneNode(node, /* deep */ true, /* withoutLoc */ true);
+    expect(cloned.loc).toBeNull();
+    expect(cloned.declarations[0].id.loc).toBeNull();
+  });
+
+  it("should support deep cloning for leadingComments, innerComments and trailingComments", function() {
+    const node = t.variableDeclaration("let", [
+      t.variableDeclarator({
+        ...t.identifier("value"),
+        typeAnnotation: t.anyTypeAnnotation(),
+      }),
+    ]);
+    node.loc = {};
+    node.declarations[0].id.loc = {};
+    const cloned = t.cloneNode(node, /* deep */ true, /* withoutLoc */ true);
+    expect(cloned.loc).toBeNull();
+    expect(cloned.declarations[0].id.loc).toBeNull();
+    expect(cloned.declarations[0].id.leadingComments).toEqual(
```

Comment:
The `cloned.declarations[0].id.leadingComments` will be `undefined` given that we did not add such comments when constructing `node`.

Similar to what we did in `node.declarations[0].id`, we can create the leading comments node
```js
node.leadingComments = [{ loc: {} }]
```
And then we assert that after it is cloned, the `loc` will be `null`:
```js
expect(cloned.leadingComments[0].loc).toBeNull()
```

Tip: you can view the node structures on [this website](https://lihautan.com/babel-ast-explorer/#?eyJiYWJlbFNldHRpbmdzIjp7InZlcnNpb24iOiI3LjYuMCIsIm9wdGlvbmFsQ2hhaW5pbmciOnsib3B0aW9ucyI6e30sImVuYWJsZWQiOnRydWV9LCJkeW5hbWljSW1wb3J0Ijp7Im9wdGlvbnMiOnt9LCJlbmFibGVkIjp0cnVlfX0sInRyZWVTZXR0aW5ncyI6eyJoaWRlRW1wdHkiOnRydWUsImhpZGVMb2NhdGlvbiI6ZmFsc2UsImhpZGVUeXBlIjp0cnVlLCJoaWRlQ29tbWVudHMiOmZhbHNlfSwiY29kZSI6Ii8vIGxlYWRpbmcgY29tbWVudHNcbmxldCB2YWx1ZSA9IDQyO1xuLy8gdHJhaWxpbmcgY29tbWVudHMifQ==), and `cloneWithoutLoc` aims to eliminate all the `loc` location information from the AST nodes.

Reply:
Thank you for the tip and the link, now I understand what I am doing!

Reply:
Thank you for the tip and the link, now I understand what I am doing!